fail when no subset solves poly system; better removal of redundant solutions

smichr opened the pull request:
<!-- Your title above should be a short description of what
was changed. Do not include the issue number in the title. -->

Until the Groebner approach is used to solve systems of polynomials, it might be advantageous to try solve the equations with the traditional methods (instead of a system-of-equations approach). The added test fails to solve as a system, but a solution can be found.

#### References to other Issues or PRs
<!-- If this pull request fixes an issue, write "Fixes #NNNN" in that exact
format, e.g. "Fixes #1234" (see
https://tinyurl.com/auto-closing for more information). Also, please
write a comment on that issue linking back to this pull request once it is
open. -->
closes #27001

#### Brief description of what is fixed or changed


#### Other comments


#### Release Notes

<!-- Write the release notes for this release below between the BEGIN and END
statements. The basic format is a bulleted list with the name of the subpackage
and the release note for this PR. For example:

* solvers
  * Added a new solver for logarithmic equations.

* functions
  * Fixed a bug with log of integers. Formerly, `log(-x)` incorrectly gave `-log(x)`.

* physics.units
  * Corrected a semantical error in the conversion between volt and statvolt which
    reported the volt as being larger than the statvolt.

or if no release note(s) should be included use:

NO ENTRY

See https://github.com/sympy/sympy/wiki/Writing-Release-Notes for more
information on how to write release notes. The bot will check your release
notes automatically to see if they are formatted correctly. -->

<!-- BEGIN RELEASE NOTES -->
* solvers
    * better removal of redundant solutions from systems of equations
    * positive-dimensional systems of polynomials which raised an error may now report a solution
<!-- END RELEASE NOTES -->

smichr committed:
fail when no subset solves polys

oscarbenjamin commented on sympy/solvers/tests/test_solvers.py:
```diff
@@ -2571,6 +2571,14 @@ def test_issue_20747():
 
 def test_issue_27001():
     assert solve((x, x**2), (x, y, z), dict=True) == [{x: 0}]
+    s = a1, a2, a3, a4, a5 = symbols('a1:6')
+    eqs = [8*a1**4*a2 + 4*a1**2*a2**3 - 8*a1**2*a2*a4 + a2**5/2 - 2*a2**3*a4 +
+        8*a2*a3**2 + 2*a2*a4**2 + 8*a2*a5, 12*a1**4 + 6*a1**2*a2**2 -
+        8*a1**2*a4 + 3*a2**4/4 - 2*a2**2*a4 + 4*a3**2 + a4**2 + 4*a5, 16*a1**3
+        + 4*a1*a2**2 - 8*a1*a4, -8*a1**2*a2 - 2*a2**3 + 4*a2*a4]
+    assert solve(eqs, s) == [
+        {a4: 2*a1**2 + a2**2/2, a5: -a3**2}, {a1: 0, a2: 0, a5: -a3**2 - a4**2/4},
+        {a1: 0, a4: a2**2/2, a5: -a3**2}]
```
With master I get:
```python
In [5]: sols = solve(groebner(eqs, s), s, dict=True)

In [6]: for s in sols: print(s)
{a4: 2*a1**2 + a2**2/2, a5: -a3**2}
{a1: 0, a2: 0, a5: -a3**2 - a4**2/4}
{a2: 0, a4: 2*a1**2, a5: -a3**2}
{a1: 0, a2: 0, a4: 0, a5: -a3**2}
{a1: 0, a2: 0, a4: 0, a5: -a3**2}
```
The 5th case is redundant and the first two match what you have here. The 4th case here is a special case of your 3rd case (with `a2=a4=0`). The 3rd case here does not seem to be included in your solution set.

smichr replied:
From the Groebner solutions, the 3rd equation is a special case of the 1st.

oscarbenjamin commented:
At some point I wrote this code:
```python
def factors_system(sets):
    """Find set of sets

    >>> factors_system([{'A', 'B'}, {'A', 'B', 'C'}])
    [{'A'}, {'B'}]
    >>> factors_system([{'A', 'B'}, {'C', 'D'}])
    [{'A', 'C'}, {'A', 'D'}, {'B', 'C'}, {'B', 'D'}]
    """
    if not sets:
        return [set()]
    elif not all(sets):
        return []

    all_items = set().union(*sets)

    A = all_items.pop()

    eqs_without_A = [eq for eq in sets if A not in eq]
    sets_with_A = [s | {A} for s in factors_system(eqs_without_A)]

    all_eqs_A_removed = [eq - {A} for eq in sets]
    sets_without_A = factors_system(all_eqs_A_removed)

    return sets_with_A + sets_without_A

def factors_system(eqs):
    """Find set of sets

    >>> factors_system([{'A', 'B'}, {'A', 'B', 'C'}])
    [{'A'}, {'B'}]
    >>> factors_system([{'A', 'B'}, {'C', 'D'}])
    [{'A', 'C'}, {'A', 'D'}, {'B', 'C'}, {'B', 'D'}]
    >>> factors_system([set('ABC'), set('BCD')])
    [{'B'}, {'A', 'D'}, {'C'}]
    """
    sets = _factors_system(eqs)
    out = []
    for s1 in sets:
        if not any(s1 > s2 for s2 in sets):
            out.append(s1)
    return out

def _factors_system(eqs):
    if not eqs:
        return [set()]
    elif len(eqs) == 1:
        return [{f} for f in eqs[0]]
    else:
        eq1, eqs_rest = eqs[0], eqs[1:]
        sets = []
        for f in eq1:
            eqs_rest_2 = [eq for eq in eqs_rest if f not in eq]
            sets.extend(s | {f} for s in factors_system(eqs_rest_2))
        return sets

def factors(*args, **kwargs):
    return list(dict(factor_list(*args, **kwargs)[1]))

def factor_poly_system(eqs):
    eqs = [set(factors(eq)) for eq in eqs]
    return factors_system(eqs)


import doctest
doctest.testmod()
```
I don't remember if that code is completely correct though because I may have been in the middle of editing it...

Its purpose is to take a system of polynomial equations and factorise them and then generate smaller systems of polynomial equations whose combined solutions sets generate all solutions of the original system. Applied to your example test case we have:
```python
In [7]: systems = factor_poly_system(eqs)

In [8]: for sys in systems:
   ...:     print()
   ...:     for eq in sys:
   ...:         print(eq)
   ...: 

16*a1**4 + 8*a1**2*a2**2 - 16*a1**2*a4 + a2**4 - 4*a2**2*a4 + 16*a3**2 + 4*a4**2 + 16*a5
48*a1**4 + 24*a1**2*a2**2 - 32*a1**2*a4 + 3*a2**4 - 8*a2**2*a4 + 16*a3**2 + 4*a4**2 + 16*a5
4*a1**2 + a2**2 - 2*a4

a1
a2
48*a1**4 + 24*a1**2*a2**2 - 32*a1**2*a4 + 3*a2**4 - 8*a2**2*a4 + 16*a3**2 + 4*a4**2 + 16*a5

48*a1**4 + 24*a1**2*a2**2 - 32*a1**2*a4 + 3*a2**4 - 8*a2**2*a4 + 16*a3**2 + 4*a4**2 + 16*a5
a2
4*a1**2 + a2**2 - 2*a4
```
So the original system of 4 equations was factorised and different combinations of the factors were then recombined to make 3 smaller polynomial systems having only 3 equations each. Regardless of anything else it is worth doing this before calling `groebner` because computing a Groebner basis can be extremely slow so we should reduce the problem as much as possible first. These three systems are also now more trivially solvable either as they are or vai a Groebner basis:
```python
In [9]: for eq in groebner(systems[0], s): print(eq)
4*a1**2 + a2**2 - 2*a4
a3**2 + a5

In [10]: for eq in groebner(systems[1], s): print(eq)
a1
a2
4*a3**2 + a4**2 + 4*a5

In [11]: for eq in groebner(systems[2], s): print(eq)
2*a1**2 - a4
a2
a3**2 + a5
```
Applying the back substitution that I described previously we can solve them. Your solutions given in the test case contain the solutions to the 1st and 2nd  cases here but not the all solutions to the 3rd which can be solved as:
```python
{a4: 2*a1**2, a2: 0, a5: -a3**2}
```
This is because the 3rd here is a special case of the first and your 3rd case is a different special case.

We can prove that one is a special case of the other by using a Groebner basis to reduce the others equations to zero:
```python
In [16]: gb = groebner(systems[2], s)

In [17]: [gb.reduce(eq)[-1] for eq in systems[0]]
Out[17]: [0, 0, 0]
```
This proves that as a system of equations `systems[2]` is a superset of `systems[0]`. That means that the solutions of `systems[0]` are a superset of those of `systems[2]`. Hence `systems[2]` is redundant and can be dropped leaving us only with the first two cases and so the solution set has two cases.

smichr commented on sympy/solvers/tests/test_solvers.py:
```diff
@@ -2571,6 +2571,14 @@ def test_issue_20747():
+        8*a2*a3**2 + 2*a2*a4**2 + 8*a2*a5, 12*a1**4 + 6*a1**2*a2**2 -
+        8*a1**2*a4 + 3*a2**4/4 - 2*a2**2*a4 + 4*a3**2 + a4**2 + 4*a5, 16*a1**3
+        + 4*a1*a2**2 - 8*a1*a4, -8*a1**2*a2 - 2*a2**3 + 4*a2*a4]
+    assert solve(eqs, s) == [
```
```suggestion
    assert solve(eqs, s, dict=True) == [
```

smichr committed:
Apply suggestions from code review

oscarbenjamin commented on sympy/solvers/tests/test_solvers.py:
```diff
@@ -2571,6 +2571,14 @@ def test_issue_20747():
+        8*a1**2*a4 + 3*a2**4/4 - 2*a2**2*a4 + 4*a3**2 + a4**2 + 4*a5, 16*a1**3
+        + 4*a1*a2**2 - 8*a1*a4, -8*a1**2*a2 - 2*a2**3 + 4*a2*a4]
+    assert solve(eqs, s, dict=True) == [
+        {a4: 2*a1**2 + a2**2/2, a5: -a3**2}, {a1: 0, a2: 0, a5: -a3**2 - a4**2/4},
+        {a1: 0, a4: a2**2/2, a5: -a3**2}]
```
There should be two cases here not three. The third case is redundant because it corresponds to setting `a1=0` in the first case.

smichr replied:
```suggestion
        {a4: 2*a1**2 + a2**2/2, a5: -a3**2}, {a1: 0, a2: 0, a5: -a3**2 - a4**2/4}]
```

smichr committed:
update existing soln

When testing to see if a solution is a superset of another, update the other with the values of the solution being tested, first.

smichr committed:
Update sympy/solvers/tests/test_solvers.py

smichr committed:
Update test_solvers.py

oscarbenjamin commented:
Okay, well let's merge this for now since it fixes the issue.
